### ISSUE TITLE
Possibility of Unity (Android, iOS, Mac OS X, Linux) Support?

### ISSUE THREAD

**Issue by corytrese** [smourier/SQLNado#6]
I'm curious if the library's authors have any information about the possibility of porting / extending / supporting this library for additional platforms under Unity / Mono.

In my case I have the sqlite3 library built for each platform (Android, iOS, Mac OS X, Linux) and I'd like to use SQLNado over the library on those OS.

Is this even remotely possible? I am interested in doing the port / contributing the bridge code if it is possible -- I would be using it in a game project, compliant with the licenses.

Any guidance, suggestions, or rough ideas / concepts are very welcome. Thanks for your time and thank you for the wonderful SQLite library!

**Comment by smourier** [comment]
Hi Cory,

I developed SQLNado using .NET standard (2+) with that exact idea in mind :-)

Basically, any platform that support .NET standard (which is .NET Framework, UWP, .NET Core, Xamarin/Mono and associated OSes) *should* be able to use SQLite provided the binary is available (which is the case for SQLite).

Every interop code is located in SQLiteDatabase.cs. I think it shouldn't be a lot of work, probably tweak DllImport values, and maybe some signature, arguments, argument types, etc...

If you're used to .NET Core on Linux, that would be the easiest platform to start with. If only I had some spare time I could try it, but my technical knowledge beyond Windows is currently weak, I must admit :-)

**Comment by corytrese** [comment]
This is wonderful news, thank you very much for the response. 

I have very little cross-platform .NET experience -- I am a C++ developer and when I have used C# it has been on Win32. What I do have is a reasonable level of expertise with building SQLite3 libraries for every platform.

I will start with .NET Core on Linux and see what I can come up with. It sounds like you've already done all the hard work!

**Comment by smourier** [comment]
Ok, great, don't hesitate to come back to me if I can do anything.

**Comment by smourier** [comment]
Note I've modified SQLNado so it's more cross-platform ready. It currently support Windows and Linux, with more to come.

**Comment by corytrese** [comment]
Let me know if I can offer any help, I have a modified version that works on Android/iOS/Mac OS X/Linux/Win64/Nintendo Switch with Unity IL2CPP.

I also have a modified version that supports some of the standard commercial extensions from SQLite like SEE, CEROD and ZIPVFS.

**Comment by smourier** [comment]
Just an update: SQLNado now supports Android fine (using SQLite official Android builds).

**Comment by corytrese** [comment]
SQLNado is really an incredible product. There are so many ways to use it. 

We found that using Android's built in library provided unacceptable performance for real time games. However, we did use it that way for some 2d turn based games for about five years.

For Unity, the best approach is to modify SQLNado to use IL2CPP plugins. This will give you real-time ready performance (up to three times faster than the default) and the option to mix in any SQLite3 add-ons like SEE or CEROD.

**Comment by smourier** [comment]
About performance in general (couldn't say in your particular context), it should be close to what SQLite provides.

Some things to make sure:

* when you do lots of updates, is much better (order of magnitude) to create a surrounding transaction when doing multiple updates. I always found that counterintuitive in comparison with other databases "transactions", but it's a game changer. This one is on SQLite.
* depending on context, you might want to change how SQLite works, with various settings, for example:

```
db = new SQLiteDatabase(...)

// https://stackoverflow.com/a/47745836/403671
db.MemoryMapSize = -1;

...

db.JournalMode = SQLiteJournalMode.Off,
db.SynchronousMode = SQLiteSynchronousMode.Off,
db.LockingMode = SQLiteLockingMode.Exclusive

...
```
* when you do updates and the schema (table/indices) doesn't change it's much better to "synchronize" tables and indices once per "session" (whatever that means in the context, can be once per process in Windows), and use an option in save calls. If you don't do that, SQLNado will try to synchronize on each save call. This one is on SQLNado. So, something like this in an hypothetical "sales" database :

```
public sealed class SalesDatabase : SQLiteDatabase
{
    private readonly SQLiteSaveOptions _saveOptions;

    public SalesDatabase(ProductsDatabase products, ILogger<SalesDatabase>? logger)
        : base(Path.Combine(FileSystem.Current.AppDataDirectory, "sales.db"))
    {
        foreach (var type in SalesTableTypes)
        {
            // so we don't have to do at each call
            SynchronizeSchema(type);
            SynchronizeIndices(type);
        }

        // we synchronize at startup, so we don't need to sync at each save
        _saveOptions = CreateSaveOptions();
        _saveOptions.SynchronizeSchema = false;
        _saveOptions.SynchronizeIndices = false;
        
        ...
    }

    public override bool Save(object obj, SQLiteSaveOptions? options = null)
    {
        // we synchronize at startup, so we don't need to sync at each save
        options ??= _saveOptions;
        return base.Save(obj, options);
    }

    // list of tables known to this database, that we want to sync only once 
    private static IEnumerable<Type> SalesTableTypes
    {
        get
        {
            yield return typeof(Order);
            yield return typeof(BasketProduct);
            yield return typeof(FavoriteProduct);
            yield return typeof(Setting);
        }
    }
}

```

**Comment by corytrese** [comment]
Great SQLite performance tips. Android gamedevs will know the performance issues I'm talking about. 

Unity gamedevs (who are compiling C# and then converting the IL bytecode to C++ and recompiling it with native plugin source mixed in) will recognize the advantages of __Internal in the IL2CPP toolchain, especially on mono runtimes.
